File: antia_numerical_methods/f77/sub4/betcon.f
!	To calculate the incomplete Beta function I_x(a,b) using
!	the continued fraction (modified form), called by BETAP
!
!	A,B : (input) Arguments for the complete Beta function
!	X : (input) Upper limit of integration defining the incomplete
!                       Beta function
!
!	Required routines : GAMMAL

      FUNCTION BETCON(A,B,X)
!      IMPLICIT REAL*8(A-H,O-Z)
      DIMENSION C(500),D(500)
           
51    FORMAT(' ** Roundoff error while evaluating the modified ',
     1  'continued fraction for incomplete Beta function at',/,'  A=',
     2  1PE12.4,'  B=',E12.4,'  X=',E12.4,'  CON. FRAC.=',E12.4)

      C(1)=A
      D(1)=A*(1-X*(A+B)/(A+1))
      B1=1+X*(B-1)/(A+1)-X*(A+1)*(A+B+1)/(A+3)
      C(2)=A*B1
      A1=A*X*X*(B-1)*(A+B)/(A+1)**2
      D(2)=B1*D(1)+A1
      C1=C(2)/D(2)
      DO I=2,499
        D1=-X*(A+I-1)*(A+B+I-1)/((A+2*I-2)*(A+2*I-1))
        D3=-X*(A+I)*(A+B+I)/((A+2*I)*(A+2*I+1))
        D2=X*I*(B-I)/((A+2*I-1)*(A+2*I))
        C(I+1)=C(I)*(A+2*I)*(1+D2+D3)-C(I-1)*(A+2*I)*(A+2*I-2)*D2*D1
        D(I+1)=D(I)*(A+2*I)*(1+D2+D3)-D(I-1)*(A+2*I)*(A+2*I-2)*D2*D1
!	Scale the numerator and denominator to prevent underflow/overflow
        IF(ABS(C(I+1)).GT.1.D30) THEN
          C(I+1)=C(I+1)/1.D30
          D(I+1)=D(I+1)/1.D30
          C(I)=C(I)/1.D30
          D(I)=D(I)/1.D30
        ENDIF
        IF(ABS(C(I+1)).LT.1.D-30) THEN
          C(I+1)=C(I+1)*1.D30
          D(I+1)=D(I+1)*1.D30
          C(I)=C(I)*1.D30
          D(I)=D(I)*1.D30
        ENDIF
        C2=C(I+1)/D(I+1)
        IF(ABS(C2-C1).LT.1.E-9) EXIT
        C1=C2
      ENDDO
      IF(C2.LT.0.0) PRINT 51,A,B,X,C2
      B1=A*LOG(X)+B*LOG(1-X)+LOG(C2)-LOG(A)
      B1=B1+GAMMAL(A+B)-GAMMAL(A)-GAMMAL(B)
      BETCON=EXP(B1)
      END

File: antia_numerical_methods/c/fun/betcon.c
/*	To calculate the incomplete Beta function I_x(a,b) using
	the continued fraction (modified form), called by betap

	a,b : (input) Arguments for the complete Beta function
	x : (input) Upper limit of integration defining the incomplete
			Beta function

	Required routines : gammaln
*/


#include <math.h>
#include <stdio.h>

double gammaln(double x);

double betcon(double a, double b, double x)
{
	double c[500],d[500],a1,b1,c1,c2,d1,d2,d3;
	int i;

	c[0]=a; d[0]=a*(1-x*(a+b)/(a+1));
	b1=1+x*(b-1)/(a+1)-x*(a+1)*(a+b+1)/(a+3);
	a1=a*x*x*(b-1)*(a+b)/pow(a+1,2.0);
	c[1]=a*b1; d[1]=b1*d[0]+a1; c2=c[1]/d[1]; c1=0.0; i=1;

	while(i<500 && fabs(c1-c2)>1.e-12)
	{
		i=i+1; c1=c2;
		d1=-x*(a+i-1)*(a+b+i-1)/((a+2*i-2)*(a+2*i-1));
		d3=-x*(a+i)*(a+b+i)/((a+2*i)*(a+2*i+1));
		d2=x*i*(b-i)/((a+2*i-1)*(a+2*i));
		c[i]=c[i-1]*(a+2*i)*(1+d2+d3)-c[i-2]*(a+2*i)*(a+2*i-2)*d2*d1;
		d[i]=d[i-1]*(a+2*i)*(1+d2+d3)-d[i-2]*(a+2*i)*(a+2*i-2)*d2*d1;
		if(fabs(c[i])>1.e200)
/*	Scale the numerator and denominator to prevent underflow/overflow */
		{
			c[i]=c[i]/1e200; c[i-1]=c[i-1]/1e200;
			d[i]=d[i]/1e200; d[i-1]=d[i-1]/1e200;
		}
		if(fabs(c[i])<1.e-200)
		{
			c[i]=c[i]*1e200; c[i-1]=c[i-1]*1e200;
			d[i]=d[i]*1e200; d[i-1]=d[i-1]*1e200;
		}
		c2=c[i]/d[i];
	}
	if(c2<0.0) printf(" ** Roundoff error while evaluating the modified continued fraction for incomplete Beta function at\n a= %e  b=%e  x= %e cont. frac.= %e\n",a,b,x,c2);
	b1=a*log(x)+b*log(1-x)+log(c2)-log(a);
	b1=b1+gammaln(a+b)-gammaln(a)-gammaln(b);
	return exp(b1);
}

